dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,662 @@
+	// Spec and SpecRef are mutually exclusive.
+	// +optional
+	Spec *ResourceClaimSpec `json:"spec"`
+
+	// SpecRef references a separate object with the specification of the claim
+	// that will be used by the driver when allocating a resource for the
+	// claim. Parameters and ParametersRef are mutually exclusive.
+	//
+	// SpecRef is typically used to reference a vendor CR.
+	// A vendor controller then converts that CR into a ResourceClaimSpecification
+	// object and that is then used to allocate the claim.
 	//
+	// The object must be in the same namespace as the ResourceClaim.
 	// +optional
+	SpecRef *ResourceClaimSpecReference `json:"specRef,omitempty"`
+}
+
+// Used inside a ResourceClaimSpecAlternatives or a ResourceClaimSpecification object.
+type ResourceClaimSpec struct {
+	// ResourceClassName references additional configuration and requirements
+	// that apply to the whole claim and all requests in it. If the class
+	// contains default requests, then those are used if (and only if)
+	// the claim does not provide those itself.
+	//
+	// Requirements in the class must be satisfied in addition to the
+	// requirements in the claim.
+	//
+	// +optional
+	ResourceClassName string `json:"resourceClassName,omitempty" protobuf:"bytes,1,name=resourceClassName"`
+
+	// To be decided: does it make sense to allow referencing multiple
+	// classes?  As it stands now, there could be a need to create a large
+	// variety of different classes where each class is one combination of
+	// different options.
+	//
+	// When allowing different classes, each class could describe one aspect:
+	// - "devices from vendor foo"
+	// - "more than 10 GiB of RAM"
+	//
+	// Probably need better use cases for classes. Stuff for a new KEP in >= 1.32...
+
+	// Config defines configuration parameters that apply to the entire claim.
+	// They are ignored while allocating the claim.
+	//
+	// +optional
+	Config *ConfigurationParameters `json:"config,omitempty" protobuf:"bytes,4,opt,name=config"`
+
+	// Requests are individual requests for separate resources for the claim.
+	// An empty list is valid and means that the claim can always be allocated
+	// without needing anything. A class can be referenced to use the default
+	// requests from that class.
+	//
 	// +listType=atomic
-	MatchAttributes []string `json:"matchAttributes,omitempty"`
+	Requests []ResourceRequest `json:"requests,omitempty" protobuf:"bytes,5,name=requests"`
 
-	// Devices contains the actual device claim details, each entry
-	// of which must be sastisfied by selecting one or more devices.
-	// Each entry is either a claim for set of devices, or a list of
-	// prioritized claims for sets of devices, one of which must be
-	// satisfied.
+	// All these match criteria must be satisfied by all devices allocated through this
+	// claim.
 	//
-	// +required
+	// +optional
 	// +listType=atomic
-	Devices []DeviceClaimInstance `json:"devices,omitempty"`
-}
+	Match []MatchModel `json:"match,omitempty"`
```

Comment:
Requests now have hard requirements. We can add new requirement types and old schedulers will fail gracefully because they see an empty one-of struct.

Here we can add more match types, but we cannot add entirely different claim requirements that aren't "a match". I think we should prepare for that and introduce a requirement list also here:

```yaml
spec:
  requirements:
  - match:
       attribute: pci-root.dra.k8s.io
  - match:
       attribute: numa.dra.k8s.io
  - someOtherFutureClaimRequirement:
       ...
```

Reply:
I went ahead and made that change, see https://github.com/kubernetes-sigs/wg-device-management/pull/14/commits/e71290b300d8ffdee0f6a95d3346c81bc4993f68.

The rationale for making `match: attribute` a single string instead of allowing a list is that we then would have two different ways to do the same thing (repeat names inside `match`, repeat `match` with single name multiple times). This would have raised the question whether it makes a difference and which one is "better".

While at it, I grouped options belonging to a claim and those belonging to a request in the ResourceClass. This is not a functional change, I Just think it reads better this way.

Reply:
I started wondering whether we should say that "claims have *constraints* and requests have *requirements*", i.e. rename `claim.requirements` -> `claim.constraints`?

Using different names makes is easier to remember what goes where. A "match attribute" is a constraint and goes into the claim, a "consume resource" is a requirement and goes into the request...

`dra-evolution/testdata/pod-two-containers-two-gpus-2.yaml`:

```yaml
apiVersion: resource.k8s.io/v1alpha2
kind: ResourceClaimTemplate
metadata:
  name: example.com-foozer-two-separate-foozer-1k
  namespace: default
spec:
  spec:
    constraints:
    - match:
        attribute: pcie-root.dra.k8s.io
    requests:
    - name: foozer-gpu
      classes:
      - name: example.com-foozer
      requirements:
      - device:
          selector: "device.stringAttributes['model.example.com-foozer'] == 'foozer-1000'"
    - name: other-foozer-gpu
      classes:
      - name: example.com-foozer
      requirements:
      - device:
          selector: "device.stringAttributes['model.example.com-foozer'] == 'foozer-1000'"
```

Reply:
This distinction and name change makes sense to me.

Reply:
Changed.